crates/biome_service/src/settings.rs

```diff
@@ -1530,10 +1511,49 @@ impl TryFrom<OverrideAssistConfiguration> for AssistSettings {
 
     fn try_from(conf: OverrideAssistConfiguration) -> Result<Self, Self::Error> {
         Ok(Self {
-            enabled: conf.enabled.unwrap_or_default(),
+            enabled: conf.enabled,
             actions: conf.actions,
             ignored_files: Matcher::empty(),
             included_files: Matcher::empty(),
         })
     }
 }
+
+/// Checks the feature activity according to language-specific
+/// and top level feature activities.
+///
+/// | Top-Level \ Language  | Some(true)   | Some(false) | None              |
+/// |:---------------------:|:------------:|:-----------:|:-----------------:|
+/// | Some(true)            | Some(true)   | Some(false) | None / Some(true) |
+/// | Some(false)           | Some(true)   | Some(false) | Some(false)       |
+/// | None                  | Some(true)   | Some(false) | None              |
+///
+/// The case `Some(true)` (top-level) / `None` (language) varies based on `is_override`.
+///
+/// - If `is_override` is `false` we don't want a top level
+///   feature to override the language-specific feature whose default
+///   value is false but in an "unset" state (`None`). So that we can
+///   still use `.unwrap_or_default()` to retrieve the correct
+///   fallback value. This happens when we want to mark the features
+///   of some languages as opt-in.
+///
+/// - If `is_override` is `true`, we don't care if a feature is meant to be opt-in, so we fall back
+///   to the top-level configuration.
+pub(crate) fn check_feature_activity<const LANG: bool, const TOP: bool>(
```

Comment:
I think we better split this function into two: One for overrides and one for non-overrides. The boolean doesn't make for great readability at the call sites (I had to jump to the implementation to understand what it even meant), and as the function documentation already indicates, the actual implementation is quite different between the two versions.